pkg/R/DataFrame.R

```diff
@@ -114,20 +114,192 @@ setMethod("count",
+# but allow for use with DataFrames by first converting to an RRDD before calling #
+# the requested map function.                                                     #
+###################################################################################
+
+setMethod("lapply",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            rdd <- toRDD(X)
+            lapply(rdd, FUN)
+          })
+
+setMethod("map",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            lapply(X, FUN)
+          })
+
+setMethod("flatMap",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            rdd <- toRDD(X)
+            flatMap(rdd, FUN)
+          })
+
+setMethod("lapplyPartition",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            rdd <- toRDD(X)
+            lapplyPartition(rdd, FUN)
+          })
+
+setMethod("mapPartitions",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            lapplyPartition(X, FUN)
+          })
+
+setMethod("lapplyPartitionsWithIndex",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            rdd <- toRDD(X)
+            lapplyPartitionsWithIndex(rdd, FUN)
+          })
+
+setMethod("mapPartitionsWithIndex",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN){
+            lapplyPartitionsWithIndex(X, FUN)
+          })
+
+setMethod("foreach",
+          signature(rdd = "DataFrame", func = "function"),
+          function(rdd, func) {
+            rddIn <- toRDD(rdd)
+            foreach(rddIn, func)
+          })
+
+setMethod("foreachPartition",
+          signature(rdd = "DataFrame", func = "function"),
+          function(rdd, func) {
+            rddIn <- toRDD(rdd)
+            foreachPartition(rddIn, func)
+          })
+
+#' Collect to DataFrame
+#' 
+#' Collects all the elements of a Spark DataFrame and coerces them into an R data.frame.
+#' 
+#' @param df A Spark DataFrame
+#' @return An R data.frame
+#' 
+#' @rdname collectToDF
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' sqlCtx <- sparkRSQL.init(sc)
+#' path <- "path/to/file.json"
+#' df <- jsonFile(sqlCtx, path)
+#' converted <- collectToDF(df)
+#' is.data.frame(converted)
+#' }
+
+setGeneric("collectToDF", function(df) { standardGeneric("collectToDF") })
+
+setMethod("collectToDF",
+          signature(df = "DataFrame"),
+          function(df) {
+            listCols <- callJStatic("edu.berkeley.cs.amplab.sparkr.SQLUtils", "dfToColRDD", df@sdf)
+            cols <- lapply(seq_along(listCols),
+                           function(colIdx) {
+                             rddCol <- listCols[[colIdx]]
+                             # Returns a list of byte arrays per partition
+                             rddRaw <- callJMethod(rddCol, "collect")
```

Comment:
I think @cafreeman measured that it was slower to do the conversion row-wise instead of column-wise on the R-side. However we could do a `collect` of the DF and do the row to column wise conversion in Scala on the driver.